### ISSUE TITLE
fileam.c: Compilation (warning-turned) error under GCC 7

### ISSUE THREAD

**Issue by d** [greenplum-db/gpdb#4499]
### Alternative headline
fileam.c:1043:11: error: 'error_caught' is used uninitialized in this function [-Werror=uninitialized]

### Greenplum version or build
Master (commit 3027102451 as of Wednesday, March 21 2018)

### OS version and uname -a
Ubuntu "rolling" 17.10
GCC version 7.2.0

### autoconf options used ( config.status --config )
```
--disable-orca --with-gssapi --enable-mapreduce --with-perl --with-libxml --with-python --disable-gpcloud --disable-pxf --enable-gpfdist --enable-debug 'CXX=ccache g++' 'CC=ccache gcc'
```
Here's the resulting `CFLAGS` if it helps you:
```
configure: using CFLAGS=-O3 -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -g -ggdb  -Werror=uninitialized
configure: using CPPFLAGS= -D_GNU_SOURCE -I/usr/include/libxml2  -I/build/install/include
configure: using LDFLAGS=  -L/build/install/lib -Wl,--as-needed
configure: using LIBS=-lbz2 -lxml2 -lrt -lgssapi_krb5 -lz -lreadline -lcrypt -ldl -lm  -lcurl
```

### Expected behavior
That it compiles

### Actual behavior
Error:
```
fileam.c: In function 'externalgettup_custom':
fileam.c:1043:11: error: 'error_caught' is used uninitialized in this function [-Werror=uninitialized]
     bool  error_caught = false;
           ^~~~~~~~~~~~
In file included from ../../../../src/include/postgres.h:53:0,
                 from fileam.c:34:
../../../../src/include/utils/elog.h:399:15: error: 'save_exception_stack' is used uninitialized in this function [-Werror=uninitialized]
   sigjmp_buf *save_exception_stack = PG_exception_stack; \
               ^~~~~~~~~~~~~~~~~~~~
../../../../src/include/utils/elog.h:399:15: note: 'save_exception_stack' was declared here
   sigjmp_buf *save_exception_stack = PG_exception_stack; \
               ^
fileam.c:1048:5: note: in expansion of macro 'PG_TRY'
     PG_TRY();
     ^~~~~~
../../../../src/include/utils/elog.h:400:25: error: 'save_context_stack' is used uninitialized in this function [-Werror=uninitialized]
   ErrorContextCallback *save_context_stack = error_context_stack; \
                         ^~~~~~~~~~~~~~~~~~
../../../../src/include/utils/elog.h:400:25: note: 'save_context_stack' was declared here
   ErrorContextCallback *save_context_stack = error_context_stack; \
                         ^
fileam.c:1048:5: note: in expansion of macro 'PG_TRY'
     PG_TRY();
     ^~~~~~
```

GCC 7 seems to have two issues with this file:
1. the local variable `error_caught`
1. the macro expanded variables `save_exception_stack`

I stared at the code for a while, and it doesn't scream "uninitialized variable" to me. There's no way the control flow could have resulted in an uninitialized read. In fact, even if the compiler is unaware of the `longjmp` and `setjmp` shenanigan, there's no chance this could be an uninitialized read, so I can't in good conscience add a `volatile` there to shut it up either.

### Step to reproduce the behavior
Compile Greenplum on latest (17.10) Ubuntu with GCC and G++ installed.
Alternatively, try configuring Greenplum on other UNIXes with GCC-7 (`configure CC=gcc-7`)

### Note
Even though the repro `configure` flags above include `--disable-orca`, it doesn't really matter. The compilation of `fileam.c` errors out regardless of whether we link with ORCA.

**Comment by d** [comment]
@andreasscherbaum @danielgustafsson I thought you two were recently dabbling with GCC 7, have you come across this error?

**Comment by asimrp** [comment]
I'm seeing this error too, with gcc 7.2 and O3.  Turning asserts on results in successful compilation with O3.

**Comment by d** [comment]
Talked to @lij55 briefly, to reiterate: *you don't necessarily need the latest Ubuntu to reproduce the warnings / errors* . You can configure the project with `CC=gcc-7` on MacOS (or any other OS where gcc-7 is available) and still reproduce

**Comment by d** [comment]
For some reason this stops failing recently. Master (as of commit a031b9c014c) is now passing when compiling with GCC 7 under the default flags, including `-O3`

**Comment by d** [comment]
A further `git bisect` revealed that the commit that fixed the compilation breakage is d65cd31edd4ffd32a824f1f0d3712b04f77d04b6 "Some fixes for externalgettup_custom() (#5391)"

**Comment by d** [comment]
Somehow this stops being an issue, partly because we merged #5391, partly because I moved onto GCC 8. I'm gonna close this issue. Although #6121 popped up after switching to GCC 8.2.0...
